HLSL -> GLSL : Incorrect variable declarations when using for loop

Hi,

I have coded this simple HLSL skinning vertex shader:

```
cbuffer _shared_camera
{
	row_major matrix V;
	row_major matrix invV;
    row_major matrix P;
	row_major matrix invP;
	float zNear;
	float zFar;
};
#define MAX_JOINTS 128
cbuffer skinningCB
{
    row_major matrix boneMatrices[MAX_JOINTS];
};

struct VertexIn
{
	float3 position : POSITION;
	float2 uv1 : TEXCOORD0;
	uint4 boneIndices : BLENDINDICES;
	float4 boneWeights : BLENDWEIGHT;
	uint instanceID : SV_InstanceID;
};
struct VertexOut
{
	float4 pos : SV_POSITION;
	float2 tex : TEXCOORD0;
	float3 posWorld : TEXCOORD1;
	float3 posView : TEXCOORD2;
};

//--------------------------------------------------------------------------------------
// VERTEX_SHADER vs_main
//--------------------------------------------------------------------------------------

VertexOut vs_main(VertexIn input)
{
	VertexOut output;
	
	output.posWorld = float3(0,0,0);
	
	for(int i = 0; i < 4; i++)
	{
		uint boneIndex = input.boneIndices[i];
		
		if(boneIndex < MAX_JOINTS)
		{
			float boneWeight = input.boneWeights[i];
			row_major matrix boneMatrix = boneMatrices[boneIndex];
			output.posWorld += mul(float4(input.position, 1), boneMatrix) * boneWeight;
		}
	}
		
	output.posView = mul(float4(output.posWorld, 1), V).xyz;
	output.pos = mul(float4(output.posView,1), P);
	output.tex = input.uv1;
	return output;
}	
```

I have translated the shader to SPIR-V without issues using glslang, but when I try to translate to GLSL using SPIRV-Cross and compile the shader on Windows platform I get the following error message:

0(63) : error C1060: incompatible types in initialization
0(63) : error C1056: invalid initialization
0(63) : error C1020: invalid operands to "<"
0(63) : error C1022: operands to "++" must be numeric

The obtained GLSL code is the following:

```
#version 420

struct VertexIn
{
    vec3 position;
    vec2 uv1;
    uvec4 boneIndices;
    vec4 boneWeights;
    uint instanceID;
};

struct VertexOut
{
    vec4 pos;
    vec2 tex;
    vec3 posWorld;
    vec3 posView;
};

layout(std140) uniform skinningCB
{
    mat4 boneMatrices[128];
} _skinningCB;

layout(std140) uniform _shared_camera
{
    mat4 V;
    mat4 invV;
    mat4 P;
    mat4 invP;
    float zNear;
    float zFar;
} __shared_camera;

layout(location = 0) in vec3 input_position;
layout(location = 1) in vec2 input_uv1;
layout(location = 2) in uvec4 input_boneIndices;
layout(location = 3) in vec4 input_boneWeights;
uniform int SPIRV_Cross_BaseInstance;
layout(location = 0) out vec2 _entryPointOutput_tex;
layout(location = 1) out vec3 _entryPointOutput_posWorld;
layout(location = 2) out vec3 _entryPointOutput_posView;

VertexIn _267;
VertexOut _268;

void main()
{
    VertexIn _234 = _267;
    _234.position = input_position;
    VertexIn _236 = _234;
    _236.uv1 = input_uv1;
    VertexIn _238 = _236;
    _238.boneIndices = input_boneIndices;
    VertexIn _240 = _238;
    _240.boneWeights = input_boneWeights;
    VertexIn _242 = _240;
    _242.instanceID = (gl_InstanceID + SPIRV_Cross_BaseInstance);
    VertexIn param = _242;
    VertexOut _244 = _268;
    _244.posWorld = vec3(0.0);
    VertexOut _271;
    for (VertexOut _269 = _244, _270 = 0; _270 < 4; _269 = _271, _270++)
    {
        uint _177 = param.boneIndices[_270];
        if (_177 < 128u)
        {
            VertexOut _248 = _269;
            _248.posWorld = _269.posWorld + vec3(((_skinningCB.boneMatrices[_177] * vec4(param.position, 1.0)) * param.boneWeights[_270]).xyz);
            _271 = _248;
        }
        else
        {
            _271 = _269;
        }
    }
    vec3 _217 = (__shared_camera.V * vec4(_269.posWorld, 1.0)).xyz;
    VertexOut _252 = _269;
    _252.posView = _217;
    vec4 _227 = __shared_camera.P * vec4(_217, 1.0);
    VertexOut _256 = _252;
    _256.pos = _227;
    VertexOut _258 = _256;
    _258.tex = param.uv1;
    gl_Position = _227;
    _entryPointOutput_tex = param.uv1;
    _entryPointOutput_posWorld = _258.posWorld;
    _entryPointOutput_posView = _217;
}
```
If you take a look into this code line:

for (VertexOut _269 = _244, _270 = 0; _270 < 4; _269 = _271, _270++)

You can see that the variable _270 has not been declared as it should be int _270 = 0

Please, could you take a look into this issue?
Thank you very much

Sorry for my late response.

I was working with a very recent master version. Just pulled from the master branch and this issue is not showing up anymore. Thanks!

Unfortunatelly I still I am getting errors of the same type in other shaders. For example, take a look a this HLSL code (this is just a part form a big shader):

```
int GetCascade(float3 fragWorldPosition)
{
	[unroll]
	for(uint cascadeIndex = 0; cascadeIndex < shadowCascadesNum; cascadeIndex++)
	{
		matrix worldToShadowMap = mul(lightVP[cascadeIndex], GetClip2TexMatrix());
		float4 fragShadowMapPos = mul(float4(fragWorldPosition, 1), worldToShadowMap);
		
		if(fragShadowMapPos.z >= 0 && fragShadowMapPos.z <= 1 &&
		max(fragShadowMapPos.x, fragShadowMapPos.y) <= 1 && 
		min(fragShadowMapPos.x, fragShadowMapPos.y) >= 0) 
			return cascadeIndex;
	}
	return -1;
}
```

SPIRV-Cross (latest master) generates the following GLSL code:

```
int GetCascade(vec3 fragWorldPosition)
{
    uint _144;
    uint _462;
    mat4 _398;
    mat4 _401;
    for (uint _4267 = 0u; _4267 < __shared_shadows.shadowCascadesNum; _462 = (uint _4272) + uint(1), _4267 = _462)
    {
        mat4 _4273;
        do
        {
            if (__shared_renderSettings.invertVCoord == 1u)
            {
                _4272 = _4267;
                _4273 = mat4(vec4(0.5, 0.0, 0.0, 0.0), vec4(0.0, -0.5, 0.0, 0.0), vec4(0.0, 0.0, 1.0, 0.0), vec4(0.5, 0.5, 0.0, 1.0));
                break;
            }
            _4272 = _4267;
            _4273 = mat4(vec4(0.5, 0.0, 0.0, 0.0), vec4(0.0, 0.5, 0.0, 0.0), vec4(0.0, 0.0, 1.0, 0.0), vec4(0.5, 0.5, 0.0, 1.0));
            break;
        } while (false);
        vec4 _434 = (_4273 * __shared_lighting.lightVP[_4272]) * vec4(fragWorldPosition, 1.0);
        if ((((_434.z >= 0.0) && (_434.z <= 1.0)) && (max(_434.x, _434.y) <= 1.0)) && (min(_434.x, _434.y) >= 0.0))
        {
            return int(_4272);
        }
    }
    return -1;
}

```

The compiler reports the following error:

0(119) : error C0000: syntax error, unexpected identifier, expecting '(' or ')' at token "_4272"
0(128) : error C0000: syntax error, unexpected reserved word "break", expecting ')' at token "break"
0(132) : error C0000: syntax error, unexpected reserved word "break", expecting ')' at token "break"

It seems it doesn´t like this line of code

 for (uint _4267 = 0u; _4267 < __shared_shadows.shadowCascadesNum; _462 = (uint _4272) + uint(1), _4267 = _462)

I believe is complaining about _4272 declaration 

If fact now I am getting errors almost in every HLSL shader SPIRV-Cross translates to GLSL.
If you want to see the whole shader let me know.

Thank you!

--- a/reference/opt/shaders/asm/frag/loop-body-dominator-continue-access.asm.frag
+++ b/reference/opt/shaders/asm/frag/loop-body-dominator-continue-access.asm.frag
@@ -1,0 +1,56 @@
+#version 450
+
+layout(binding = 0, std140) uniform Foo
+{
+    layout(row_major) mat4 lightVP[64];
+    uint shadowCascadesNum;
+    int test;
+} _11;
+
+layout(location = 0) in vec3 fragWorld;
+layout(location = 0) out int _entryPointOutput;
+
+mat4 _152;
+uint _155;
+
+int GetCascade(vec3 fragWorldPosition)
+{
+    uint _49;
+    uint _119;
+    mat4 _43;
+    mat4 _47;
+    mat4 _153;
+    _153 = _152;
+    uint _156;
+    mat4 _157;
+    for (uint _151 = 0u; _151 < _11.shadowCascadesNum; _119 = _156 + uint(1), _151 = _119, _153 = _157)
+    {
+        mat4 _154;
+        _154 = _153;
+        for (;;)
+        {
+            if (_11.test == 0)
+            {
+                _156 = _151;
+                _157 = mat4(vec4(0.5, 0.0, 0.0, 0.0), vec4(0.0, 0.5, 0.0, 0.0), vec4(0.0, 0.0, 0.5, 0.0), vec4(0.0, 0.0, 0.0, 1.0));
+                break;
+            }
+            _156 = _151;
+            _157 = mat4(vec4(1.0, 0.0, 0.0, 0.0), vec4(0.0, 1.0, 0.0, 0.0), vec4(0.0, 0.0, 1.0, 0.0), vec4(0.0, 0.0, 0.0, 1.0));
+            break;
+        }
+        vec4 _92 = (_157 * _11.lightVP[_156]) * vec4(fragWorldPosition, 1.0);
+        if ((((_92.z >= 0.0) && (_92.z <= 1.0)) && (max(_92.x, _92.y) <= 1.0)) && (min(_92.x, _92.y) >= 0.0))
+        {
+            return int(_156);
+        }
+    }
+    return -1;
+}
+
+void main()
+{
+    vec3 _123 = fragWorld;
+    _entryPointOutput = GetCascade(_123);
+}
+

--- a/reference/shaders/asm/frag/loop-body-dominator-continue-access.asm.frag
+++ b/reference/shaders/asm/frag/loop-body-dominator-continue-access.asm.frag
@@ -1,0 +1,48 @@
+#version 450
+
+layout(binding = 0, std140) uniform Foo
+{
+    layout(row_major) mat4 lightVP[64];
+    uint shadowCascadesNum;
+    int test;
+} _11;
+
+layout(location = 0) in vec3 fragWorld;
+layout(location = 0) out int _entryPointOutput;
+
+mat4 GetClip2TexMatrix()
+{
+    if (_11.test == 0)
+    {
+        return mat4(vec4(0.5, 0.0, 0.0, 0.0), vec4(0.0, 0.5, 0.0, 0.0), vec4(0.0, 0.0, 0.5, 0.0), vec4(0.0, 0.0, 0.0, 1.0));
+    }
+    return mat4(vec4(1.0, 0.0, 0.0, 0.0), vec4(0.0, 1.0, 0.0, 0.0), vec4(0.0, 0.0, 1.0, 0.0), vec4(0.0, 0.0, 0.0, 1.0));
+}
+
+int GetCascade(vec3 fragWorldPosition)
+{
+    for (uint cascadeIndex = 0u; cascadeIndex < _11.shadowCascadesNum; cascadeIndex++)
+    {
+        mat4 worldToShadowMap = GetClip2TexMatrix() * _11.lightVP[cascadeIndex];
+        vec4 fragShadowMapPos = worldToShadowMap * vec4(fragWorldPosition, 1.0);
+        if ((((fragShadowMapPos.z >= 0.0) && (fragShadowMapPos.z <= 1.0)) && (max(fragShadowMapPos.x, fragShadowMapPos.y) <= 1.0)) && (min(fragShadowMapPos.x, fragShadowMapPos.y) >= 0.0))
+        {
+            return int(cascadeIndex);
+        }
+    }
+    return -1;
+}
+
+int _main(vec3 fragWorld_1)
+{
+    vec3 param = fragWorld_1;
+    return GetCascade(param);
+}
+
+void main()
+{
+    vec3 fragWorld_1 = fragWorld;
+    vec3 param = fragWorld_1;
+    _entryPointOutput = _main(param);
+}
+

--- a/shaders/asm/frag/loop-body-dominator-continue-access.asm.frag
+++ b/shaders/asm/frag/loop-body-dominator-continue-access.asm.frag
@@ -1,0 +1,190 @@
+; SPIR-V
+; Version: 1.0
+; Generator: Khronos Glslang Reference Front End; 2
+; Bound: 131
+; Schema: 0
+               OpCapability Shader
+          %1 = OpExtInstImport "GLSL.std.450"
+               OpMemoryModel Logical GLSL450
+               OpEntryPoint Fragment %main "main" %fragWorld_1 %_entryPointOutput
+               OpExecutionMode %main OriginUpperLeft
+               OpSource HLSL 500
+               OpName %main "main"
+               OpName %GetClip2TexMatrix_ "GetClip2TexMatrix("
+               OpName %GetCascade_vf3_ "GetCascade(vf3;"
+               OpName %fragWorldPosition "fragWorldPosition"
+               OpName %_main_vf3_ "@main(vf3;"
+               OpName %fragWorld "fragWorld"
+               OpName %Foo "Foo"
+               OpMemberName %Foo 0 "lightVP"
+               OpMemberName %Foo 1 "shadowCascadesNum"
+               OpMemberName %Foo 2 "test"
+               OpName %_ ""
+               OpName %cascadeIndex "cascadeIndex"
+               OpName %worldToShadowMap "worldToShadowMap"
+               OpName %fragShadowMapPos "fragShadowMapPos"
+               OpName %param "param"
+               OpName %fragWorld_0 "fragWorld"
+               OpName %fragWorld_1 "fragWorld"
+               OpName %_entryPointOutput "@entryPointOutput"
+               OpName %param_0 "param"
+               OpDecorate %_arr_mat4v4float_uint_64 ArrayStride 64
+               OpMemberDecorate %Foo 0 RowMajor
+               OpMemberDecorate %Foo 0 Offset 0
+               OpMemberDecorate %Foo 0 MatrixStride 16
+               OpMemberDecorate %Foo 1 Offset 4096
+               OpMemberDecorate %Foo 2 Offset 4100
+               OpDecorate %Foo Block
+               OpDecorate %_ DescriptorSet 0
+               OpDecorate %_ Binding 0
+               OpDecorate %fragWorld_1 Location 0
+               OpDecorate %_entryPointOutput Location 0
+       %void = OpTypeVoid
+          %3 = OpTypeFunction %void
+      %float = OpTypeFloat 32
+    %v4float = OpTypeVector %float 4
+%mat4v4float = OpTypeMatrix %v4float 4
+          %9 = OpTypeFunction %mat4v4float
+    %v3float = OpTypeVector %float 3
+%_ptr_Function_v3float = OpTypePointer Function %v3float
+        %int = OpTypeInt 32 1
+         %15 = OpTypeFunction %int %_ptr_Function_v3float
+       %uint = OpTypeInt 32 0
+    %uint_64 = OpConstant %uint 64
+%_arr_mat4v4float_uint_64 = OpTypeArray %mat4v4float %uint_64
+        %Foo = OpTypeStruct %_arr_mat4v4float_uint_64 %uint %int
+%_ptr_Uniform_Foo = OpTypePointer Uniform %Foo
+          %_ = OpVariable %_ptr_Uniform_Foo Uniform
+      %int_2 = OpConstant %int 2
+%_ptr_Uniform_int = OpTypePointer Uniform %int
+      %int_0 = OpConstant %int 0
+       %bool = OpTypeBool
+  %float_0_5 = OpConstant %float 0.5
+    %float_0 = OpConstant %float 0
+         %39 = OpConstantComposite %v4float %float_0_5 %float_0 %float_0 %float_0
+         %40 = OpConstantComposite %v4float %float_0 %float_0_5 %float_0 %float_0
+         %41 = OpConstantComposite %v4float %float_0 %float_0 %float_0_5 %float_0
+    %float_1 = OpConstant %float 1
+         %43 = OpConstantComposite %v4float %float_0 %float_0 %float_0 %float_1
+         %44 = OpConstantComposite %mat4v4float %39 %40 %41 %43
+         %46 = OpConstantComposite %v4float %float_1 %float_0 %float_0 %float_0
+         %47 = OpConstantComposite %v4float %float_0 %float_1 %float_0 %float_0
+         %48 = OpConstantComposite %v4float %float_0 %float_0 %float_1 %float_0
+         %49 = OpConstantComposite %mat4v4float %46 %47 %48 %43
+%_ptr_Function_uint = OpTypePointer Function %uint
+     %uint_0 = OpConstant %uint 0
+      %int_1 = OpConstant %int 1
+%_ptr_Uniform_uint = OpTypePointer Uniform %uint
+%_ptr_Function_mat4v4float = OpTypePointer Function %mat4v4float
+%_ptr_Uniform_mat4v4float = OpTypePointer Uniform %mat4v4float
+%_ptr_Function_v4float = OpTypePointer Function %v4float
+     %uint_2 = OpConstant %uint 2
+%_ptr_Function_float = OpTypePointer Function %float
+     %uint_1 = OpConstant %uint 1
+     %int_n1 = OpConstant %int -1
+%_ptr_Input_v3float = OpTypePointer Input %v3float
+%fragWorld_1 = OpVariable %_ptr_Input_v3float Input
+%_ptr_Output_int = OpTypePointer Output %int
+%_entryPointOutput = OpVariable %_ptr_Output_int Output
+       %main = OpFunction %void None %3
+          %5 = OpLabel
+%fragWorld_0 = OpVariable %_ptr_Function_v3float Function
+    %param_0 = OpVariable %_ptr_Function_v3float Function
+        %125 = OpLoad %v3float %fragWorld_1
+               OpStore %fragWorld_0 %125
+        %129 = OpLoad %v3float %fragWorld_0
+               OpStore %param_0 %129
+        %130 = OpFunctionCall %int %_main_vf3_ %param_0
+               OpStore %_entryPointOutput %130
+               OpReturn
+               OpFunctionEnd
+%GetClip2TexMatrix_ = OpFunction %mat4v4float None %9
+         %11 = OpLabel
+         %30 = OpAccessChain %_ptr_Uniform_int %_ %int_2
+         %31 = OpLoad %int %30
+         %34 = OpIEqual %bool %31 %int_0
+               OpSelectionMerge %36 None
+               OpBranchConditional %34 %35 %36
+         %35 = OpLabel
+               OpReturnValue %44
+         %36 = OpLabel
+               OpReturnValue %49
+               OpFunctionEnd
+%GetCascade_vf3_ = OpFunction %int None %15
+%fragWorldPosition = OpFunctionParameter %_ptr_Function_v3float
+         %18 = OpLabel
+%cascadeIndex = OpVariable %_ptr_Function_uint Function
+%worldToShadowMap = OpVariable %_ptr_Function_mat4v4float Function
+%fragShadowMapPos = OpVariable %_ptr_Function_v4float Function
+               OpStore %cascadeIndex %uint_0
+               OpBranch %55
+         %55 = OpLabel
+               OpLoopMerge %57 %58 Unroll
+               OpBranch %59
+         %59 = OpLabel
+         %60 = OpLoad %uint %cascadeIndex
+         %63 = OpAccessChain %_ptr_Uniform_uint %_ %int_1
+         %64 = OpLoad %uint %63
+         %65 = OpULessThan %bool %60 %64
+               OpBranchConditional %65 %56 %57
+         %56 = OpLabel
+         %68 = OpFunctionCall %mat4v4float %GetClip2TexMatrix_
+         %69 = OpLoad %uint %cascadeIndex
+         %71 = OpAccessChain %_ptr_Uniform_mat4v4float %_ %int_0 %69
+         %72 = OpLoad %mat4v4float %71
+         %73 = OpMatrixTimesMatrix %mat4v4float %68 %72
+               OpStore %worldToShadowMap %73
+         %76 = OpLoad %mat4v4float %worldToShadowMap
+         %77 = OpLoad %v3float %fragWorldPosition
+         %78 = OpCompositeExtract %float %77 0
+         %79 = OpCompositeExtract %float %77 1
+         %80 = OpCompositeExtract %float %77 2
+         %81 = OpCompositeConstruct %v4float %78 %79 %80 %float_1
+         %82 = OpMatrixTimesVector %v4float %76 %81
+               OpStore %fragShadowMapPos %82
+         %85 = OpAccessChain %_ptr_Function_float %fragShadowMapPos %uint_2
+         %86 = OpLoad %float %85
+         %87 = OpFOrdGreaterThanEqual %bool %86 %float_0
+         %88 = OpAccessChain %_ptr_Function_float %fragShadowMapPos %uint_2
+         %89 = OpLoad %float %88
+         %90 = OpFOrdLessThanEqual %bool %89 %float_1
+         %91 = OpLogicalAnd %bool %87 %90
+         %92 = OpAccessChain %_ptr_Function_float %fragShadowMapPos %uint_0
+         %93 = OpLoad %float %92
+         %95 = OpAccessChain %_ptr_Function_float %fragShadowMapPos %uint_1
+         %96 = OpLoad %float %95
+         %97 = OpExtInst %float %1 FMax %93 %96
+         %98 = OpFOrdLessThanEqual %bool %97 %float_1
+         %99 = OpLogicalAnd %bool %91 %98
+        %100 = OpAccessChain %_ptr_Function_float %fragShadowMapPos %uint_0
+        %101 = OpLoad %float %100
+        %102 = OpAccessChain %_ptr_Function_float %fragShadowMapPos %uint_1
+        %103 = OpLoad %float %102
+        %104 = OpExtInst %float %1 FMin %101 %103
+        %105 = OpFOrdGreaterThanEqual %bool %104 %float_0
+        %106 = OpLogicalAnd %bool %99 %105
+               OpSelectionMerge %108 None
+               OpBranchConditional %106 %107 %108
+        %107 = OpLabel
+        %109 = OpLoad %uint %cascadeIndex
+        %110 = OpBitcast %int %109
+               OpReturnValue %110
+        %108 = OpLabel
+               OpBranch %58
+         %58 = OpLabel
+        %112 = OpLoad %uint %cascadeIndex
+        %113 = OpIAdd %uint %112 %int_1
+               OpStore %cascadeIndex %113
+               OpBranch %55
+         %57 = OpLabel
+               OpReturnValue %int_n1
+               OpFunctionEnd
+ %_main_vf3_ = OpFunction %int None %15
+  %fragWorld = OpFunctionParameter %_ptr_Function_v3float
+         %21 = OpLabel
+      %param = OpVariable %_ptr_Function_v3float Function
+        %118 = OpLoad %v3float %fragWorld
+               OpStore %param %118
+        %119 = OpFunctionCall %int %GetCascade_vf3_ %param
+               OpReturnValue %119
+               OpFunctionEnd

--- a/spirv_cross.cpp
+++ b/spirv_cross.cpp
@@ -1976,6 +1976,8 @@ void Compiler::parse(const Instruction &instruction)
 		loop_blocks.insert(current_block->self);
 		loop_merge_targets.insert(current_block->merge_block);
 
+		continue_block_to_loop_header[current_block->continue_block] = current_block->self;
+
 		// Don't add loop headers to continue blocks,
 		// which would make it impossible branch into the loop header since
 		// they are treated as continues.
@@ -3214,6 +3216,10 @@ void Compiler::analyze_variable_scope(SPIRFunction &entry)
 				auto *var = compiler.maybe_get<SPIRVariable>(ptr);
 				if (var && var->storage == StorageClassFunction)
 					accessed_variables_to_block[var->self].insert(current_block->self);
+
+				for (uint32_t i = 3; i < length; i++)
+					if (id_is_phi_variable(args[i]))
+						accessed_variables_to_block[args[i]].insert(current_block->self);
 				break;
 			}
 
@@ -3352,16 +3358,27 @@ void Compiler::analyze_variable_scope(SPIRFunction &entry)
 		{
 			// If we're accessing a variable inside a continue block, this variable might be a loop variable.
 			// We can only use loop variables with scalars, as we cannot track static expressions for vectors.
-			if (this->is_continue(block) && type.vecsize == 1 && type.columns == 1)
+			if (this->is_continue(block))
 			{
-				// The variable is used in multiple continue blocks, this is not a loop
-				// candidate, signal that by setting block to -1u.
-				auto &potential = potential_loop_variables[var.first];
+				// Potentially awkward case to check for.
+				// We might have a variable inside a loop, which is touched by the continue block,
+				// but is not actually a loop variable.
+				// The continue block is dominated by the inner part of the loop, which does not make sense in high-level
+				// language output because it will be declared before the body,
+				// so we will have to lift the dominator up to the relevant loop header instead.
+				builder.add_block(continue_block_to_loop_header[block]);
 
-				if (potential == 0)
-					potential = block;
-				else
-					potential = ~(0u);
+				if (type.vecsize == 1 && type.columns == 1)
+				{
+					// The variable is used in multiple continue blocks, this is not a loop
+					// candidate, signal that by setting block to -1u.
+					auto &potential = potential_loop_variables[var.first];
+
+					if (potential == 0)
+						potential = block;
+					else
+						potential = ~(0u);
+				}
 			}
 			builder.add_block(block);
 		}
@@ -3370,6 +3387,7 @@ void Compiler::analyze_variable_scope(SPIRFunction &entry)
 
 		// Add it to a per-block list of variables.
 		uint32_t dominating_block = builder.get_dominator();
+
 		// If all blocks here are dead code, this will be 0, so the variable in question
 		// will be completely eliminated.
 		if (dominating_block)

--- a/spirv_cross.hpp
+++ b/spirv_cross.hpp
@@ -511,6 +511,7 @@ protected:
 	std::unordered_set<uint32_t> loop_merge_targets;
 	std::unordered_set<uint32_t> selection_merge_targets;
 	std::unordered_set<uint32_t> multiselect_merge_targets;
+	std::unordered_map<uint32_t, uint32_t> continue_block_to_loop_header;
 
 	virtual std::string to_name(uint32_t id, bool allow_alias = true) const;
 	bool is_builtin_variable(const SPIRVariable &var) const;

--- a/spirv_glsl.cpp
+++ b/spirv_glsl.cpp
@@ -7846,6 +7846,9 @@ bool CompilerGLSL::attempt_emit_loop_header(SPIRBlock &block, SPIRBlock::Method 
 			{
 			case SPIRBlock::ForLoop:
 			{
+				// This block may be a dominating block, so make sure we flush undeclared variables before building the for loop header.
+				flush_undeclared_variables(block);
+
 				// Important that we do this in this order because
 				// emitting the continue block can invalidate the condition expression.
 				auto initializer = emit_for_loop_initializers(block);
@@ -7856,6 +7859,8 @@ bool CompilerGLSL::attempt_emit_loop_header(SPIRBlock &block, SPIRBlock::Method 
 			}
 
 			case SPIRBlock::WhileLoop:
+				// This block may be a dominating block, so make sure we flush undeclared variables before building the while loop header.
+				flush_undeclared_variables(block);
 				statement("while (", to_expression(block.condition), ")");
 				break;
 

